File: app/channels/conversation_channel.rb

class ConversationChannel < ApplicationCable::Channel
  def subscribed
    conversation = load_conversation

    unless conversation.includes_user?(current_user)
      reject
      return
    end

    stream_for conversation
  end

  def unsubscribed
    # no-op
  end

  # optional: allow sending from client (we usually prefer HTTP POST to create)
  def speak(data)
    conversation = load_conversation
    Message.create!(conversation:, user: current_user, body: data["body"].to_s)
  end

  # ---- helpers ----
  private

  def load_conversation
    Conversation.find(params[:conversation_id])
  end

  # class-level broadcaster helper used by model callback
  class << self
    def broadcast_message(message)
      conversation = message.conversation
      html = ApplicationController.render(
        partial: "messages/message",
        formats: [:html],
        locals: { message: }
      )
      broadcast_to(conversation, { type: "message", html:, id: message.id })
    end
  end
end

File: app/javascript/channels/conversation_channel.js

import consumer from "./consumer"

function mountConversationChannel() {
  const list = document.getElementById("messages")
  if (!list) return

  const conversationId = list.dataset.conversationId

  consumer.subscriptions.create(
    { channel: "ConversationChannel", conversation_id: conversationId },
    {
      connected() {
        // console.log("Connected to conversation", conversationId)
      },
      disconnected() {},
      received(data) {
        if (data.type === "message" && data.html) {
          list.insertAdjacentHTML("beforeend", data.html)
          list.lastElementChild?.scrollIntoView({ behavior: "smooth", block: "end" })
        }
      }
    }
  )
}

// Support Turbo navigation
document.addEventListener("turbo:load", mountConversationChannel)
document.addEventListener("turbo:render", mountConversationChannel)


